docs/haskell-use-cases.rst

```diff
@@ -143,6 +143,231 @@ This works for any ``haskell_binary`` or ``haskell_library`` target.
 Modules of all libraries will be loaded in interpreted mode and can be
 reloaded using the ``:r`` GHCi command when source files change.
 
+Configuring IDE integration with ghcide
+---------------------------------------
+
+rules_haskell has preliminary support for IDE integration using `ghcide`_. The
+ghcide project provides IDE features for Haskell projects through the Language
+Server Protocol. To set this up you can define a `haskell_repl`_ target that
+will collect the required compiler flags for your Haskell targets and pass them
+to `hie-bios`_ which will then forward them to ghcide.
+
+Let's set this up for the following example project::
+
+  haskell_toolchain_library(
+      name = "base",
+  )
+
+  haskell_library(
+      name = "library-a",
+      srcs = ["Lib/A.hs"],
+      deps = [":base"],
+  )
+
+  haskell_library(
+      name = "library-b",
+      srcs = ["Lib/B.hs"],
+      deps = [":base"],
+  )
+
+  haskell_binary(
+      name = "binary",
+      srcs = ["Main.hs"],
+      deps = [
+          ":base",
+          ":library-a",
+          ":library-b",
+      ],
+  )
+
+We want to configure ghcide to provide IDE integration for all these three
+targets. Start by defining a ``haskell_repl`` target as follows::
+
+  haskell_repl(
+    name = "hie-bios",
+    collect_data = False,
+    deps = [
+      ":binary",
+      # ":library-a",
+      # ":library-b",
+    ],
+  )
+
+Note, that ``library-a`` and ``library-b`` do not have to be listed explicitly.
+By default haskell_repl will include all transitive dependencies that are not
+external dependencies. Refer to the API documentation of `haskell_repl`_ for
+details.
+
+We also disable building runtime dependencies using ``collect_data = False`` as
+they are not required for an IDE session.
+
+You can test if this provides the expected compiler flags by running the
+following Bazel command and taking a look at the generated file::
+
+  bazel build //:hie-bios --output_groups=hie_bios
+
+Next, we need to hook this up to `hie-bios`_ using the `bios cradle`_. To that
+end, define a small shell script named ``.hie-bios`` that looks as follows::
+
+  #!/usr/bin/env bash
+  set -euo pipefail
+  bazel build //:hie-bios --output_groups=hie_bios
+  cat bazel-bin/hie-bios@hie-bios >"$HIE_BIOS_OUTPUT"
```

Comment:
I'm not sure where to put this, but I ran into some issues with targets building with `-Werror` because ghcide forcibly sets some warnings (like [missing-signatures](https://github.com/digital-asset/ghcide/issues/77)), so the combination of both might lead to spurious errors. The common way out of this is that most hie bioses seem to filter out `-Werror` or add `-Wwarn`, and that's what I also did by added an `echo -Wwarn` at the end of `.hie_bios`

Reply:
That's a good point, another way to deal with this is to add `-Wwarn` to the toolchain's GHCi flags, e.g. https://github.com/digital-asset/daml/blob/a47c734401a70e5b9dd94cd9717a2192bc28c703/WORKSPACE#L382.

Reply:
Nvm, I see that the hie-bios output group doesn't include those. Though maybe it should.